### PR TITLE
Fix fulfillments refund with waiting for approval state

### PR DESCRIPTION
**Pull request by tomaszszymanski129**
I want to merge this change because it fixes fulfillments waiting for approval not properly handled on refund option.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by tomaszszymanski129** [comment]
Re-worked the logic, now works also for partial refunds
- fulfillment lines are analyzed
- if line is waiting for approval, remove it and add quantity to order lines input of refund logic
- if quantity is lower than total - don't delete, lower the waiting fulfillment line quantity accordingly and add "to refund" quanity to order line as above

that way we use exactly same refund mechanism as before but treat "to refund" waiting amounts as part of order line unfulfilled quantity - waiting fulfillements in fact are unfulfilled in terms of warehouses and allocations.

**Comment by korycins** [comment]
@tomaszszymanski129  Will a re-worked logic fix a case when I want to proceed with a refund and I see in the refund tab, only unfulfilled section, no matter how many fulfillments `waiting for approving` I have?

**Comment by tomaszszymanski129** [comment]
@korycins frontend is now adjusted to display waiting ones separately, you can easily decide what exactly you wish to refund. Yes, works also for a mix of waiting / unfulfilled.

File: saleor/graphql/order/mutations/fulfillments.py
Comment (korycins) [inline saleor/graphql/order/mutations/fulfillments.py:730]: Not sure of this (or maybe I am missing something).  Why do we have to do that - `cls.remove_waiting_fulfillments` and `cls.extend_order_lines`?
Wouldn't be good to treat fulfillment `Waiting for Approving` as a fulfillment without allocated stock, and extend a logic inside `create_refund_fulfillment` to properly handle it (especially `saleor.order.actions._move_fulfillment_lines_to_target_fulfillment`)?  
It should work by default.

Comment (tomaszszymanski129) [reply saleor/graphql/order/mutations/fulfillments.py:730]: That was my frist approach but there is already way more complicated logic there, which also touches stocks and stuff based on quantity allocated (which doesn't include waiting ones), it was hard to get it to even work and calculate things properly, plus it made a mess with extra `if`s in already complicated places, thats why it's now more "chained".

Comment (korycins) [reply saleor/graphql/order/mutations/fulfillments.py:730]: I understand that we don't want to allocate lines from `waiting for approving`, but is it OK that we don't update quantity_fulfilled after we move lines to the fulfillment - `waiting for approving`?
Personally, I would try to extend `_move_fulfillment_lines_to_target_fulfillment` even if this will include some ifs or complicates the logic.  This should also solve the issue with return/replace case. In your case `return/replace` will also require such additional actions. For me, it looks like we change the data in order and in fulfillments to fit in old logic for refunds. 
Also, the assumption of `create_refund_fulfillment` is that you can call it and it will cover for you all things related to refund, but it will not work well with the new fulfillment type.

**Review by maarcingebala — APPROVED** [review]
I think this code should be refactored as @korycins suggests, but for now we can merge it, as it needs to be released soon.

**Pull request merged**
